File: app/src/main/java/com/example/jacob/android_sprint1_challenge/MainActivity.java
package com.example.jacob.android_sprint1_challenge;

import android.app.Activity;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.graphics.Paint;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    private Context context;
    private LinearLayout layoutList;
    private MovieViewModel viewModel;

    public static final int EDIT_REQUEST_CODE = 2;
    public static final int DELETE_REQUEST_CODE = 3;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        context = this;
        layoutList = findViewById(R.id.layout_list);

        viewModel = ViewModelProviders.of(this).get(MovieViewModel.class);

        final Observer<ArrayList<Movie>> observer = new Observer<ArrayList<Movie>>() {
            @Override
            public void onChanged(@Nullable ArrayList<Movie> movies) {
                if (movies != null) {
                    refreshListView(movies);
                }

            }
        };
        viewModel.getMovieList().observe(this, observer);

        findViewById(R.id.button_addmovie).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(context, EditActivity.class);
                Movie newMovie = null;
                intent.putExtra(EditActivity.EDIT_MOVIE_KEY, newMovie);
                startActivityForResult(intent, EDIT_REQUEST_CODE);
            }
        });
    }

    private TextView getDefaultTextView(final Movie movie) {
        TextView textView = new TextView(context);
        textView.setText(movie.getTitle());
        boolean watched = movie.getWatched();
        if (watched == true) {
            textView.setPaintFlags(textView.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
        }
        textView.setTextSize(24);
        textView.setPadding(10, 10, 10, 10);
        textView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(context, EditActivity.class);
                intent.putExtra(EditActivity.EDIT_MOVIE_KEY, movie);
                startActivityForResult(intent, EDIT_REQUEST_CODE);
            }
        });

        return textView;
    }

    private void refreshListView(ArrayList<Movie> movies) {
        layoutList.removeAllViews();
        for (Movie movie : movies) {
            layoutList.addView(getDefaultTextView(movie));
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == EDIT_REQUEST_CODE) {
                if (data != null) {
                    Movie returnedMovie = (Movie) data.getSerializableExtra(EditActivity.EDIT_MOVIE_KEY);
                    viewModel.addMovie(returnedMovie);
                }
            }
        }  else if (resultCode == Constants.DELETE_RESULT_CODE) {
            if (data != null) {
                Movie returnedMovie = (Movie) data.getSerializableExtra(EditActivity.EDIT_MOVIE_KEY);
                viewModel.deleteMovie(returnedMovie);
            }
        }
    }
}

File: app/src/main/java/com/example/android_sprint_final/MainActivity.kt
package com.example.android_sprint_final

import android.app.Activity
import android.content.Intent
import android.graphics.Paint
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.LinearLayout
import android.widget.TextView
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    companion object{
        const val MOVIE_REQUEST = 654
        const val TEXT_VIEW_REQUEST = "143"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        movie_button.setOnClickListener {view ->
            val intent = Intent(this, EditActivity::class.java)
            startActivityForResult(intent, MOVIE_REQUEST)
        }
    }

    fun createTextView(content : String, has_been_watched : Boolean) : TextView{
        val text_view = TextView(this)
        text_view.text = content
        text_view.textSize = 24f
        text_view.layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)

        if(has_been_watched == true){
            text_view.setPaintFlags(Paint.STRIKE_THRU_TEXT_FLAG)
        }

        text_view.setOnClickListener {
            movie_layout.removeView(text_view)
            val intent = Intent(this, EditActivity::class.java)
            intent.putExtra(TEXT_VIEW_REQUEST, text_view.getText().toString())
            startActivityForResult(intent, MOVIE_REQUEST)
        }

        return text_view
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if(resultCode == Activity.RESULT_OK){
            when(requestCode){
                MOVIE_REQUEST -> {
                    val text = data?.getStringExtra(EditActivity.MOVIE_ENTRY)
                    val watched = data?.getBooleanExtra(EditActivity.WATCHED_OR_NOT, false)
                    val delete = data?.getStringExtra(EditActivity.DELETE_MOVIE)

                    if(text != null && text.equals("") == false && watched != null){
                        val t : TextView = createTextView(text, watched)
                        movie_layout.addView(t)
                    } else if(delete != null && !delete.equals("")){
                        for (x in 0..(movie_layout.childCount)){
                            val t_view : TextView = movie_layout.getChildAt(x) as TextView
                            if(t_view.text.equals(delete)){
                                movie_layout.removeViewAt(x)
                            }
                        }
                    }
                }
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }
}

